src/elf/symver.rs

```diff
@@ -59,876 +59,866 @@
-    ///
-    /// https://refspecs.linuxbase.org/LSB_5.0.0/LSB-Core-generic/LSB-Core-generic/symversion.html#VERDEFENTRIES
-    #[repr(C)]
-    #[derive(Debug, Pread)]
-    struct ElfVerdef {
-        vd_version: u16,
-        vd_flags: u16,
-        vd_ndx: u16,
-        vd_cnt: u16,
-        vd_hash: u32,
-        vd_aux: u32,
-        vd_next: u32,
-    }
-
-    /// An ELF `Version Definition Auxiliary` entry Elfxx_Verdaux.
-    ///
-    /// https://refspecs.linuxbase.org/LSB_5.0.0/LSB-Core-generic/LSB-Core-generic/symversion.html#VERDEFEXTS
-    #[repr(C)]
-    #[derive(Debug, Pread)]
-    struct ElfVerdaux {
-        vda_name: u32,
-        vda_next: u32,
-    }
-
-    /// An ELF `Version Need` entry Elfxx_Verneed.
-    ///
-    /// https://refspecs.linuxbase.org/LSB_5.0.0/LSB-Core-generic/LSB-Core-generic/symversion.html#VERNEEDFIG
-    #[repr(C)]
-    #[derive(Debug, Pread)]
-    struct ElfVerneed {
-        vn_version: u16,
-        vn_cnt: u16,
-        vn_file: u32,
-        vn_aux: u32,
-        vn_next: u32,
-    }
-
-    /// An ELF `Version Need Auxiliary` entry Elfxx_Vernaux.
-    ///
-    /// https://refspecs.linuxbase.org/LSB_5.0.0/LSB-Core-generic/LSB-Core-generic/symversion.html#VERNEEDEXTFIG
-    #[repr(C)]
-    #[derive(Debug, Pread)]
-    struct ElfVernaux {
-        vna_hash: u32,
-        vna_flags: u16,
-        vna_other: u16,
-        vna_name: u32,
-        vna_next: u32,
-    }
-
-    /********************
-     *  Symbol Version  *
-     ********************/
-
-    /// Helper struct to iterate over [Symbol Version][Versym] entries.
-    #[derive(Debug)]
-    pub struct VersymSection<'a> {
+use crate::container;
+use crate::elf::section_header::{SectionHeader, SHT_GNU_VERDEF, SHT_GNU_VERNEED, SHT_GNU_VERSYM};
+use crate::error::{Error, Result};
+use crate::strtab::Strtab;
+use core::iter::FusedIterator;
+use scroll::Pread;
+
+/******************
+ *  ELF Constants *
+ ******************/
+
+// Versym constants.
+
+/// Constant describing a local symbol, see [`Versym::is_local`].
+pub const VER_NDX_LOCAL: u16 = 0;
+/// Constant describing a global symbol, see [`Versym::is_global`].
+pub const VER_NDX_GLOBAL: u16 = 1;
+/// Bitmask to check hidden bit, see [`Versym::is_hidden`].
+pub const VERSYM_HIDDEN: u16 = 0x8000;
+/// Bitmask to get version information, see [`Versym::version`].
+pub const VERSYM_VERSION: u16 = 0x7fff;
+
+// Verdef constants.
+
+/// Bitmask to check `base` flag in [`Verdef::vd_flags`].
+pub const VER_FLG_BASE: u16 = 0x1;
+/// Bitmask to check `weak` flag in [`Verdef::vd_flags`].
+pub const VER_FLG_WEAK: u16 = 0x2;
+/// Bitmask to check `info` flag in [`Verdef::vd_flags`].
+pub const VER_FLG_INFO: u16 = 0x4;
+
+/********************
+ *  ELF Structures  *
+ ********************/
+
+/// An ELF `Symbol Version` entry.
+///
+/// https://refspecs.linuxbase.org/LSB_5.0.0/LSB-Core-generic/LSB-Core-generic/symversion.html#SYMVERTBL
+#[repr(C)]
+#[derive(Debug, Pread)]
+struct ElfVersym {
```

Comment:
Alternatively, since this appears identical to Versym, and I don't think the size is unified over 32/64 bit forms, it might be fine to just have repr(C) and Pread on Versym? But this is also fine for consistency with the other types , so it's also fine :) (also i don't know much about Versym, but I don't think we'd extend it in the future for some reason with fields other than vs_val: u16, so maybe it is better future proof to have this internal and Versym public as you have it now. :shrug: